@xrstf github doesn't let me comment on code that hasn't change but when I execute script I have an error when it try to show the report. otherwise the works good :)

in the functionn `show_cover_report` surronding `$2` with quote generate error

```bash
show_cover_report() {
  echo "Generating $1 report"
  go tool cover -"$1"="$profile" "$2"
}

```

call `show_cover_report func`: 
```bash
Generating func report
+ go tool cover -func=.cover/cover.out ''
too many arguments
```

call `show_cover_report html "-o $workdir/cover.html"`:
```
Generating html report
+ go tool cover -html=.cover/cover.out '-o .cover/cover.html'
flag provided but not defined: -o .cover/cover.html
Usage of 'go tool cover':
Given a coverage profile produced by 'go test':
	go test -coverprofile=c.out

Open a web browser displaying annotated source code:
	go tool cover -html=c.out

Write out an HTML file instead of launching a web browser:
	go tool cover -html=c.out -o coverage.html

Display coverage percentages to stdout for each function:
	go tool cover -func=c.out

Finally, to generate modified source code with coverage annotations
(what go test -cover does):
	go tool cover -mode=set -var=CoverageVariableName program.go

Flags:
  -V	print version and exit
  -func string
    	output coverage profile information for each function
  -html string
    	generate HTML representation of coverage profile
  -mode string
    	coverage mode: set, count, atomic
  -o string
    	file for output; default: stdout
  -var string
    	name of coverage variable to generate (default "GoCover")

  Only one of -html, -func, or -mode may be set.
make: *** [cover] Error 1
```

possibles fixes:
1. remove the quote around `$2` `show_cover_report` in but I guess shellcheck will not like it
2. use a `if` in   `show_cover_report` 
```bash
show_cover_report() {
  echo "Generating $1 report"
  if [ -z "$2" ]; then
    go tool cover -"$1"="$profile"
  else
     go tool cover -"$1"="$profile" -o "$2"
  fi
}
```
and change  call to HTML report from `show_cover_report html "-o $workdir/cover.html" ` to `show_cover_report html "$workdir/cover.html"`

--- a/Makefile
+++ b/Makefile
@@ -141,7 +141,7 @@ spellcheck:
 
 .PHONY: cover
 cover:
-	./hack/cover.sh --html
+	./hack/coverage.sh --html
 
 .PHONY: run-controller-manager
 run-controller-manager:

--- a/hack/README.md
+++ b/hack/README.md
@@ -13,10 +13,7 @@ Create a changelog since last release, commit and create a new release tag
 
 Generate test coverage statistics for Go packages.
 
-Works around the fact that `go test -coverprofile` currently does not work
-with multiple packages, see https://code.google.com/p/go/issues/detail?id=6909
-
-    Usage: cover.sh [--html]
+    Usage: coverage.sh [--html]
 
     --html      Additionally create HTML report and open it in browser
 
@@ -77,6 +74,10 @@ to prewarm a local Docker registry, for example in offline setups.
 
 TBD
 
+## run-api-tests.sh
+
+TBD
+
 ## run-ccm-migration-e2e-test-in-kind.sh
 
 This script sets up a local KKP installation in kind, deploys a
@@ -109,14 +110,6 @@ In this mode, you need to have the kube-test binaries and all other
 dependencies installed locally on your machine, but it makes testing
 against a local KKP setup much easier.
 
-## run-crd-migration-in-kind.sh
-
-This script sets up two kind clusters, places a whole lot of KKP
-resources in various states into it and then runs the crd migration
-function of the KKP installer.
-Note that this script does install KKP itself, as KKP needs to be
-completely shutdown during the migration anyway.
-
 ## run-dashboard-and-api.sh
 
 TBD
@@ -126,6 +119,10 @@ TBD
 This script sets up a local KKP installation in kind, deploys a
 couple of test Presets and Users and then runs the e2e tests for the
 nodeport-proxy.
+
+## run-integration-tests.sh
+
+TBD
 
 ## run-machine-controller.sh
 
@@ -217,6 +214,10 @@ TBD
 TBD
 
 ## verify-grafana-dashboards.sh
+
+TBD
+
+## verify-import-order.sh
 
 TBD
 

--- a/hack/ci/README.md
+++ b/hack/ci/README.md
@@ -79,6 +79,11 @@ This script sets up a local KKP installation in kind, deploys a
 couple of test Presets and Users and then runs the e2e tests for the
 API.
 
+## run-cilium-e2e-test.sh
+
+This script is used as a postsubmit job and updates the dev master
+cluster after every commit to master.
+
 ## run-conformance-tests.sh
 
 After having set up a local KKP installation, this script is then
@@ -94,6 +99,16 @@ Kubernetes conformance.
 
 This script sets up a local KKP installation in kind, deploys a
 couple of test Presets and Users and then runs the etcd-launcher tests.
+
+## run-konnectivity-e2e-test.sh
+
+This script is used as a postsubmit job and updates the dev master
+cluster after every commit to master.
+
+## run-mla-e2e-tests.sh
+
+This script sets up a local KKP installation in kind, deploys a
+couple of test Presets and Users and then runs the MLA e2e tests.
 
 ## run-offline-test.sh
 
@@ -117,7 +132,41 @@ version for the user-ssh-key-agent.
 
 TBD
 
+## setup-kubermatic-backups-in-kind.sh
+
+This script creates a local kind cluster, compiles the KKP binaries,
+creates all Docker images and loads them into the kind cluster,
+generates a custom CA and a certificate for minio to be used as backup location,
+then installs KKP using the KKP installer + operator and sets up a
+single shared master/seed system.
+This serves as the precursor for all other tests.
+
+This script should be sourced, not called, so callers get the variables
+it sets.
+
 ## setup-kubermatic-in-kind.sh
+
+This script creates a local kind cluster, compiles the KKP binaries,
+creates all Docker images and loads them into the kind cluster,
+then installs KKP using the KKP installer + operator and sets up a
+single shared master/seed system.
+This serves as the precursor for all other tests.
+
+This script should be sourced, not called, so callers get the variables
+it sets.
+
+## setup-kubermatic-mla-in-kind.sh
+
+This script creates a local kind cluster, compiles the KKP binaries,
+creates all Docker images and loads them into the kind cluster,
+then installs KKP using the KKP installer + operator and sets up a
+single shared master/seed system.
+This serves as the precursor for all other tests.
+
+This script should be sourced, not called, so callers get the variables
+it sets.
+
+## setup-mla.sh
 
 This script creates a local kind cluster, compiles the KKP binaries,
 creates all Docker images and loads them into the kind cluster,
@@ -158,6 +207,12 @@ files, like CRD examples and the Prometheus runbook.
 Runs as a postsubmit and refreshes the gocache by downloading the
 previous version, compiling everything and then tar'ing up the
 Go cache again.
+
+## verify.sh
+
+This script is used as a presubmit to check that Helm chart versions
+have been updated if charts have been modified. Without the Prow env
+vars, this script won't run properly.
 
 ## verify-user-cluster-prometheus-configs.sh
 

--- a/hack/coverage.sh
+++ b/hack/coverage.sh
@@ -16,10 +16,7 @@
 
 ### Generate test coverage statistics for Go packages.
 ###
-### Works around the fact that `go test -coverprofile` currently does not work
-### with multiple packages, see https://code.google.com/p/go/issues/detail?id=6909
-###
-###     Usage: cover.sh [--html]
+###     Usage: coverage.sh [--html]
 ###
 ###     --html      Additionally create HTML report and open it in browser
 
@@ -27,33 +24,28 @@ set -e
 
 workdir=.cover
 profile="$workdir/cover.out"
-mode="set"
+covermode="set"
 
-generate_cover_data() {
-  rm -rf "$workdir"
-  mkdir "$workdir"
+create_cover_report() {
+  local kind="$1"
+  shift
 
-  for pkg in "$@"; do
-    f="$workdir/$(echo "$pkg" | tr / -).cover"
-    go test -covermode="$mode" -coverprofile="$f" "$pkg"
-  done
-
-  echo "mode: $mode" > "$profile"
-  grep -h -v "^mode:" "$workdir"/*.cover >> "$profile"
+  echo "Generating $kind report…"
+  go tool cover "-$kind=$profile" $@
 }
 
-show_cover_report() {
-  echo "Generating $1 report"
-  go tool cover -"$1"="$profile" "$2"
-}
+rm -rf "$workdir"
+mkdir "$workdir"
 
-generate_cover_data "$(go list ./cmd/... ./pkg/...)"
-show_cover_report func
+echo "Generating coverage data…"
+go test -covermode="$covermode" -coverprofile="$profile" ./cmd/... ./pkg/...
+
+create_cover_report func
 
 case "$1" in
 "") ;;
 --html)
-  show_cover_report html "-o $workdir/cover.html"
+  create_cover_report html -o $workdir/cover.html
   ;;
 *)
   echo >&2 "error: invalid option: $1"

